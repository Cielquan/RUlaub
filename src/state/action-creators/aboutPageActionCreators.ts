import { Dispatch } from "redux";

import { AboutPageActionType } from "../action-types";
import { AboutPageAction } from "../actions";

export const openAboutPageAction = (): AboutPageAction => ({
  type: AboutPageActionType.OPEN,
});

export const openAboutPage =
  () =>
  (dispatch: Dispatch<AboutPageAction>): void => {
    dispatch(openAboutPageAction());
  };

export const closeAboutPageAction = (): AboutPageAction => ({
  type: AboutPageActionType.CLOSE,
});

export const closeAboutPage =
  () =>
  (dispatch: Dispatch<AboutPageAction>): void => {
    dispatch(closeAboutPageAction());
  };
