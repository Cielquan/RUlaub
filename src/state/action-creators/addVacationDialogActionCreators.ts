import { Dispatch } from "redux";

import { AddVacationDialogActionType } from "../action-types";
import { AddVacationDialogAction } from "../actions";

export const openAddVacationDialogAction = (): AddVacationDialogAction => ({
  type: AddVacationDialogActionType.OPEN,
});

export const openAddVacationDialog =
  () =>
  (dispatch: Dispatch<AddVacationDialogAction>): void => {
    dispatch(openAddVacationDialogAction());
  };

export const closeAddVacationDialogAction = (): AddVacationDialogAction => ({
  type: AddVacationDialogActionType.CLOSE,
});

export const closeAddVacationDialog =
  () =>
  (dispatch: Dispatch<AddVacationDialogAction>): void => {
    dispatch(closeAddVacationDialogAction());
  };
