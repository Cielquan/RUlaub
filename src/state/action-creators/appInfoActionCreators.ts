import { Dispatch } from "redux";
import { AppInfoType } from "../action-types";
import { AppInfoAction } from "../actions";

export const changeAppInfoName =
  (newName: string) =>
  (dispatch: Dispatch<AppInfoAction>): void => {
    dispatch({
      type: AppInfoType.NAME,
      payload: newName,
    });
  };

export const changeAppInfoVersion =
  (newVersion: string) =>
  (dispatch: Dispatch<AppInfoAction>): void => {
    dispatch({
      type: AppInfoType.VERSION,
      payload: newVersion,
    });
  };

export const changeAppInfoInfoText =
  (newInfoText: string) =>
  (dispatch: Dispatch<AppInfoAction>): void => {
    dispatch({
      type: AppInfoType.INFO_TEXT,
      payload: newInfoText,
    });
  };
