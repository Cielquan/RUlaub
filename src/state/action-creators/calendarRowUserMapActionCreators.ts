import { Dispatch } from "redux";
import { UsersDataSchema } from "../../types/usersData.schema";

import { CalendarRowUserMapActionType } from "../action-types";
import { CalendarRowUserMapAction } from "../actions";

export const updateCalendarRowUserMapAction = (
  payload: UsersDataSchema
): CalendarRowUserMapAction => ({
  type: CalendarRowUserMapActionType.UPDATE,
  payload,
});

export const updateCalendarRowUserMap =
  (payload: UsersDataSchema) =>
  (dispatch: Dispatch<CalendarRowUserMapAction>): void => {
    dispatch(updateCalendarRowUserMapAction(payload));
  };
