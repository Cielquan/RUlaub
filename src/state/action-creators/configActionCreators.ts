import { invoke } from "@tauri-apps/api";
import { open, save } from "@tauri-apps/api/dialog";
import { ProviderContext } from "notistack";
import { dirname } from "path";
import { Dispatch } from "redux";

import getErrorCatalogueMsg from "../../backendAPI/errorMsgCatalogue";
import {
  ConfigFileSchema as ConfigFile,
  LogLevel,
  SupportedLanguages,
  SupportedThemes,
} from "../../backendAPI/types/configFile.schema";
import { validateConfig } from "../../backendAPI/validation";
import { enqueuePersistendErrSnackbar } from "../../utils/snackbarUtils";
import { ConfigActionType } from "../action-types";
import { ConfigAction } from "../actions";

export const updateConfigAction = (payload: ConfigFile): ConfigAction => ({
  type: ConfigActionType.UPDATE,
  payload,
});

export const loadConfig =
  () =>
  async (dispatch: Dispatch<ConfigAction>): Promise<void> => {
    const data = await invoke("get_config_state");

    let conf;
    try {
      conf = await validateConfig(data);
    } catch (err) {
      invoke("log_error", {
        target: "config",
        message: `Config data validation failed: ${err}`,
        location: "state/action-creators/configActionCreators.ts-loadConfig",
      });
      return;
    }
    dispatch(updateConfigAction(conf));
  };

export const setLanguage =
  (lang: SupportedLanguages, snackbarHandles: ProviderContext) =>
  async (dispatch: Dispatch<ConfigAction>): Promise<void> => {
    let data;
    try {
      data = await invoke<ConfigFile>("set_langauge", { lang });
    } catch (err) {
      enqueuePersistendErrSnackbar(getErrorCatalogueMsg(err as string), snackbarHandles);
      return;
    }

    let conf;
    try {
      conf = await validateConfig(data);
    } catch (err) {
      invoke("log_error", {
        target: "config",
        message: `Config data validation failed: ${err}`,
        location: "state/action-creators/configActionCreators.ts-setLanguage",
      });
      return;
    }
    dispatch(updateConfigAction(conf));
  };

export const setLogLevel =
  (level: LogLevel, snackbarHandles: ProviderContext) =>
  async (dispatch: Dispatch<ConfigAction>): Promise<void> => {
    let data;
    try {
      data = await invoke<ConfigFile>("set_log_level", { level });
    } catch (err) {
      enqueuePersistendErrSnackbar(getErrorCatalogueMsg(err as string), snackbarHandles);
      return;
    }

    let conf;
    try {
      conf = await validateConfig(data);
    } catch (err) {
      invoke("log_error", {
        target: "config",
        message: `Config data validation failed: ${err}`,
        location: "state/action-creators/configActionCreators.ts-setLogLevel",
      });
      return;
    }
    dispatch(updateConfigAction(conf));
  };

export const setTheme =
  (theme: SupportedThemes, snackbarHandles: ProviderContext) =>
  async (dispatch: Dispatch<ConfigAction>): Promise<void> => {
    let data;
    try {
      data = await invoke<ConfigFile>("set_theme", { theme });
    } catch (err) {
      enqueuePersistendErrSnackbar(getErrorCatalogueMsg(err as string), snackbarHandles);
      return;
    }

    let conf;
    try {
      conf = await validateConfig(data);
    } catch (err) {
      invoke("log_error", {
        target: "config",
        message: `Config data validation failed: ${err}`,
        location: "state/action-creators/configActionCreators.ts-setTheme",
      });
      return;
    }
    dispatch(updateConfigAction(conf));
  };

export const setTodayAutoscrollLeftOffset =
  (offset: number, snackbarHandles: ProviderContext) =>
  async (dispatch: Dispatch<ConfigAction>): Promise<void> => {
    let data;
    try {
      data = await invoke<ConfigFile>("set_today_autoscroll_left_offset", {
        offset,
      });
    } catch (err) {
      enqueuePersistendErrSnackbar(getErrorCatalogueMsg(err as string), snackbarHandles);
      return;
    }

    let conf;
    try {
      conf = await validateConfig(data);
    } catch (err) {
      invoke("log_error", {
        target: "config",
        message: `Config data validation failed: ${err}`,
        location: "state/action-creators/configActionCreators.ts-setTodayAutoscrollLeftOffset",
      });
      return;
    }
    dispatch(updateConfigAction(conf));
  };

export const setUserName =
  (name: string, snackbarHandles: ProviderContext) =>
  async (dispatch: Dispatch<ConfigAction>): Promise<void> => {
    let data;
    try {
      data = await invoke<ConfigFile>("set_user_name", { name });
    } catch (err) {
      enqueuePersistendErrSnackbar(getErrorCatalogueMsg(err as string), snackbarHandles);
      return;
    }

    let conf;
    try {
      conf = await validateConfig(data);
    } catch (err) {
      invoke("log_error", {
        target: "config",
        message: `Config data validation failed: ${err}`,
        location: "state/action-creators/configActionCreators.ts-setUserName",
      });
      return;
    }
    dispatch(updateConfigAction(conf));
  };

export const setYearChangeScrollBegin =
  (doScroll: boolean, snackbarHandles: ProviderContext) =>
  async (dispatch: Dispatch<ConfigAction>): Promise<void> => {
    let data;
    try {
      data = await invoke<ConfigFile>("set_year_change_scroll_begin", { doScroll });
    } catch (err) {
      enqueuePersistendErrSnackbar(getErrorCatalogueMsg(err as string), snackbarHandles);
      return;
    }

    let conf;
    try {
      conf = await validateConfig(data);
    } catch (err) {
      invoke("log_error", {
        target: "config",
        message: `Config data validation failed: ${err}`,
        location: "state/action-creators/configActionCreators.ts-setYearChangeScrollBegin",
      });
      return;
    }
    dispatch(updateConfigAction(conf));
  };

export const setYearToShow =
  (year: number, snackbarHandles: ProviderContext) =>
  async (dispatch: Dispatch<ConfigAction>): Promise<void> => {
    let data;
    try {
      data = await invoke<ConfigFile>("set_year_to_show", { year });
    } catch (err) {
      enqueuePersistendErrSnackbar(getErrorCatalogueMsg(err as string), snackbarHandles);
      return;
    }

    let conf;
    try {
      conf = await validateConfig(data);
    } catch (err) {
      invoke("log_error", {
        target: "config",
        message: `Config data validation failed: ${err}`,
        location: "state/action-creators/configActionCreators.ts-setYearToShow",
      });
      return;
    }
    dispatch(updateConfigAction(conf));
  };

const FILTERS = [{ name: "Database", extensions: ["db"] }];

export const createNewDB =
  (snackbarHandles: ProviderContext) =>
  async (dispatch: Dispatch<ConfigAction>): Promise<void> => {
    const path = await save({ filters: FILTERS });
    if (path === null) return;

    let data;
    try {
      data = await invoke<ConfigFile>("create_db", { path });
    } catch (err) {
      enqueuePersistendErrSnackbar(getErrorCatalogueMsg(err as string), snackbarHandles);
      return;
    }

    let conf;
    try {
      conf = await validateConfig(data);
    } catch (err) {
      invoke("log_error", {
        target: "config",
        message: `Config data validation failed: ${err}`,
        location: "state/action-creators/configActionCreators.ts-createNewDB",
      });
      return;
    }
    dispatch(updateConfigAction(conf));
  };

export const selectDB =
  (snackbarHandles: ProviderContext) =>
  async (dispatch: Dispatch<ConfigAction>): Promise<void> => {
    const dbUri = (await invoke<ConfigFile>("get_config_state")).settings.databaseUri;

    const path = (await open({
      defaultPath: dbUri ? dirname(dbUri) : undefined,
      filters: FILTERS,
      multiple: false,
      directory: false,
    })) as string;
    if (path === null) return;

    let data;
    try {
      data = await invoke<ConfigFile>("set_db_uri", { path });
    } catch (err) {
      enqueuePersistendErrSnackbar(getErrorCatalogueMsg(err as string), snackbarHandles);
      return;
    }

    let conf;
    try {
      conf = await validateConfig(data);
    } catch (err) {
      invoke("log_error", {
        target: "config",
        message: `Config data validation failed: ${err}`,
        location: "state/action-creators/configActionCreators.ts-selectDB",
      });
      return;
    }
    dispatch(updateConfigAction(conf));
  };
