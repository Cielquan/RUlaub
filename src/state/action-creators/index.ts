export * from "./configActionCreators";
export * from "./infoPageActionCreators";
export * from "./publicHolidaysDataActionCreators";
export * from "./publicHolidaysDialogActionCreators";
export * from "./schoolHolidaysDataActionCreators";
export * from "./settingsDialogActionCreators";
export * from "./sideMenuActionCreators";
export * from "./vacationDataActionCreators";
