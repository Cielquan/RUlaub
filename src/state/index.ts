export * as actionCreators from "./action-creators";
export { default as store } from "./store";
export * from "./reducers";
