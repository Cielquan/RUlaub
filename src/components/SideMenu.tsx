import { t } from "@lingui/macro";
import {
  Add as AddIcon,
  Build as BuildIcon,
  ChevronLeft as ChevronLeftIcon,
  Create as CreateIcon,
  Flight as FlightIcon,
  Folder as FolderIcon,
  Info as InfoIcon,
  Settings as SettingsIcon,
  Storage as StorageIcon,
} from "@mui/icons-material";
import { Divider, Drawer, IconButton, List } from "@mui/material";
import { styled } from "@mui/material/styles";
import { Box } from "@mui/system";
import React, { ReactElement } from "react";
import { useDispatch, useSelector } from "react-redux";
import { bindActionCreators } from "redux";

import { actionCreators, State } from "../state";

import SideMenuButton, {
  ButtonIcon,
  ButtonListKey,
  ButtonOnClick,
  ButtonSxStyle,
  ButtonText,
} from "./SideMenuButton";
import SideMenuSectionButton, { SectionItemList } from "./SideMenuSectionButton";

const SideMenuHeader = styled("div")(({ theme }) => ({
  width: 240,
  display: "flex",
  alignItems: "center",
  justifyContent: "flex-end",
  padding: theme.spacing(0, 1),

  // necessary to match toolbar height
  ...theme.mixins.toolbar,
}));

type SectionlessItemList = Array<
  [ButtonText, ButtonListKey, ButtonOnClick, ButtonIcon?, ButtonSxStyle?]
>;

interface Props {
  onClick?: () => void | null;
}

const SideMenu = ({ onClick }: Props): ReactElement => {
  const dispatch = useDispatch();
  const { closeSideMenu, openInfoPage } = bindActionCreators(actionCreators, dispatch);
  const sideMenuState = useSelector((state: State) => state.sideMenu);

  const sectionlessItemList: SectionlessItemList = [
    [t`Settings`, "Settings", closeSideMenu, <SettingsIcon />],
  ];

  const VacationSectionItemList: SectionItemList = [
    [t`Create new`, "Vac-Create new", closeSideMenu, <AddIcon />],
    [t`Edit existing`, "Vac-Edit existing", closeSideMenu, <CreateIcon />],
  ];

  const DatabaseSectionItemList: SectionItemList = [
    [t`Create new`, "DB-Create new", closeSideMenu, <AddIcon />],
    [t`Select existing`, "DB-Select existing", closeSideMenu, <FolderIcon />],
    [t`Modify selected`, "DB-Modify selected", closeSideMenu, <BuildIcon />],
  ];

  return (
    <Drawer
      data-testid="side-menu"
      anchor="left"
      open={sideMenuState}
      onClose={closeSideMenu}
    >
      <SideMenuHeader>
        <IconButton
          onClick={() => {
            if (typeof onClick === "function") onClick();
            closeSideMenu();
          }}
        >
          <ChevronLeftIcon />
        </IconButton>
      </SideMenuHeader>
      <Divider />
      <List sx={{ paddingTop: 0 }}>
        {sectionlessItemList.map((item) => (
          <React.Fragment key={item[1]}>
            <SideMenuButton
              text={item[0]}
              listKey={item[1]}
              onClick={() => {
                if (typeof onClick === "function") onClick();
                closeSideMenu();
                item[2]();
              }}
              icon={item[3]}
              sxStyle={item[4]}
            />
            <Divider />
          </React.Fragment>
        ))}
        <SideMenuSectionButton
          text={t`Vacation`}
          listKey="Vacation"
          icon={<FlightIcon />}
          sectionItemList={VacationSectionItemList}
        />
        <Divider />
        <SideMenuSectionButton
          text={t`Database`}
          listKey="Database"
          icon={<StorageIcon />}
          sectionItemList={DatabaseSectionItemList}
        />
      </List>
      <List
        sx={{
          marginTop: "auto",
          marginBottom: 2,
        }}
      >
        <Divider />
        <SideMenuButton
          text={t`Info`}
          listKey="Info"
          icon={<InfoIcon />}
          onClick={() => {
            if (typeof onClick === "function") onClick();
            closeSideMenu();
            openInfoPage();
          }}
        />
      </List>
    </Drawer>
  );
};
SideMenu.defaultProps = {
  onClick: () => undefined,
};

export default SideMenu;
