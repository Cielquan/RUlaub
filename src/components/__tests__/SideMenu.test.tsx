import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";
import { Provider } from "react-redux";
import { CombinedState } from "redux";
import { MockStoreEnhanced } from "redux-mock-store";

import SideMenu from "../SideMenu";
import { initialState, State } from "../../state";
import generateMockStore from "../../testUtils";
import { closeSideMenuAction } from "../../state/action-creators";

jest.mock("../SideMenuButton", () => () => "SideMenuButton");
jest.mock("../SideMenuSectionButton", () => () => "SideMenuSectionButton");

describe("<SideMenu />", () => {
  let testState: State;
  let mockStore: MockStoreEnhanced<Partial<CombinedState<State>>>;

  beforeEach(() => {
    testState = JSON.parse(JSON.stringify(initialState));
    mockStore = generateMockStore(testState);
  });

  it("renders not by default", () => {
    expect(mockStore.getState().sideMenu).toBe(false);
    render(
      <Provider store={mockStore}>
        <SideMenu />
      </Provider>
    );
    expect(screen.queryByTestId("side-menu")).not.toBeInTheDocument();
  });

  it("renders correctly when open", () => {
    testState.sideMenu = true;
    const alteredMockStore = generateMockStore(testState);
    expect(alteredMockStore.getState().sideMenu).toBe(true);
    render(
      <Provider store={mockStore}>
        <SideMenu />
      </Provider>
    );
    expect(screen.getByTestId("side-menu")).toMatchSnapshot();
  });

  it("closes when button is clicked", () => {
    testState.sideMenu = true;
    const alteredMockStore = generateMockStore(testState);
    const onClickMock = jest.fn();

    expect(alteredMockStore.getState().sideMenu).toBe(true);
    render(
      <Provider store={alteredMockStore}>
        <SideMenu onClick={onClickMock} />
      </Provider>
    );
    expect(screen.getByTestId("side-menu")).toBeVisible();

    userEvent.click(screen.getByTestId("side-menu-btn"));
    expect(onClickMock).toHaveBeenCalledTimes(1);
    const expectedActions = [closeSideMenuAction()];
    expect(alteredMockStore.getActions()).toEqual(expectedActions);
  });
});
