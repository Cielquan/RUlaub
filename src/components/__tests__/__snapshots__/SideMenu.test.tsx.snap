// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SideMenu /> renders correctly when open 1`] = `
<div
  class="MuiModal-root MuiDrawer-root MuiDrawer-modal css-1u2w381-MuiModal-root-MuiDrawer-root"
  data-testid="side-menu"
  role="presentation"
>
  <div
    aria-hidden="true"
    class="MuiBackdrop-root css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  />
  <div
    data-test="sentinelStart"
    tabindex="0"
  />
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation16 MuiDrawer-paper MuiDrawer-paperAnchorLeft css-4t3x6l-MuiPaper-root-MuiDrawer-paper"
    tabindex="-1"
  >
    <div
      class="css-1nmeykx"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ChevronLeftIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.41 7.41 14 6l-6 6 6 6 1.41-1.41L10.83 12z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
    <ul
      class="MuiList-root MuiList-padding css-1ccm1cv-MuiList-root"
    >
      SideMenuButton
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
      SideMenuSectionButton
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
      SideMenuSectionButton
    </ul>
    <ul
      class="MuiList-root MuiList-padding css-mes3qp-MuiList-root"
    >
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
      SideMenuButton
    </ul>
  </div>
  <div
    data-test="sentinelEnd"
    tabindex="0"
  />
</div>
`;
